src/Microsoft.AspNet.Mvc.Razor/RazorViewEngine.cs

```diff
@@ -240,47 +241,65 @@ internal static string GetNormalizedRouteValue(ActionContext context, string key
             }
 
             // 2. With the values that we've accumumlated so far, check if we have a cached result.
-            var pageLocation = _viewLocationCache.Get(expanderContext);
-            if (!string.IsNullOrEmpty(pageLocation))
+            var cachedResult = _viewLocationCache.Get(expanderContext);
+            IEnumerable<string> locationsToSearch = null;
+
+            if (cachedResult != null)
             {
-                var page = _pageFactory.CreateInstance(pageLocation);
+                if (cachedResult.IsFoundResult)
+                {
+                    var page = _pageFactory.CreateInstance(cachedResult.ViewLocation);
```

Comment:
We could look at all the views we previously looked at to see if they exist now. This would resolve the issue where a view with a better location appears since we last cached it.